ceph/rados/serviceability_workflows.py

```diff
@@ -315,6 +315,125 @@ def remove_osds_from_host(self, host_obj):
                     service_type="osd", service_name=service
                 )
 
+    def unmanaged_osd_service_exists(self) -> bool:
+        """
+        Method to check if the cluster has unmanaged service based OSDs
+        Args:
+        Returns:
+            unmanaged service based OSDs exist -> True
+            unmanaged service based OSDs do not exist -> False
+        """
+        out = self.rados_obj.list_orch_services(service_type="osd", export=True)
+        for _service in out:
+            if not (
+                _service.get("placement", False) or _service.get("service_id", False)
+            ):
+                log.info("Service %s not managed", _service)
+                return True
+        log.info("All OSD daemons are managed by named service")
+        return False
+
+    def get_cephadm_managable_osd_specs(self) -> list:
+        """
+        Method to get the valid OSD specs that are present in the cluster.
+        Args:
+        returns:
+            List of names of OSD specs that are valid
+        """
+        return [
+            name["service_id"]
+            for name in self.rados_obj.list_orch_services(
+                service_type="osd", export=True
+            )
+            if ("placement" and "service_id") in name
+        ]
+
+    def get_osd_spec(self, osd_id) -> str:
+        """
+        method to get the name of spec, to which a particular OSD belongs
+        Args:
+            osd_id: OSD ID whose spec is needed
+        Returns:
+            name of the spec
+        """
+        osd_metadata = self.rados_obj.get_daemon_metadata(
+            daemon_type="osd", daemon_id=osd_id
+        )
+        return osd_metadata["osdspec_affinity"]
+
+    def add_osds_to_managed_service(self, osds: list = None, spec: str = None) -> bool:
+        """
+        Method to identify non-managed OSDs and then add them back to managed state under existing OSD specs
+        Args:
+            osds: list of OSDs to be set to managed state
+            spec: name of the spec that needs to be used for setting OSD to manage.
+        Returns:
+            Pass -> True
+            fail -> False
+
+        Note : if osds list is provided, Only those osds will be added to managed and check if the cluster,
+        is free of unmanaged OSDs.
+
+        Note : If no OSDs are provided, we add all OSDs to managed host by host,
+        until we're left with no unmanaged osds on cluster.
+
+        Note: If Spec is not provided, 1st available - managed OSD spec name would be used
+        """
+        if not float(self.rhbuild.split("-")[0]) >= 8.1:
+            log.info(
+                "Feature not backport-ed on versions lower than 8.1. Skipping workflow"
+                "Tests running on build : %s",
+                self.rhbuild,
+            )
+            return True
+
+        if not self.unmanaged_osd_service_exists():
+            log.info("No unmanaged OSDs exist on the cluster on the start of test")
```

Comment:
Should we return True here?

Reply:
But we are claiming that no OSDs are unmanaged, so if all the OSDs are already managed, why do we need to do anything at all?